### PR TITLE
Allow safetensors offload

### PR DESCRIPTION
**Pull request by sgugger**
This PR reworks a bit the internals of the offload hooks to allow for the case where the checkpoint is saved with `safetensors`. In this case, there is no need to have a separate save of the weight as a Numpy memory-mapped array, as we can directly access the weight using `safe_open`.

Goes along with https://github.com/huggingface/transformers/pull/20321

### CODE REVIEW

File: src/accelerate/utils/offload.py [hunk 1]
```diff
@@ -152,12 +153,25 @@ def __init__(
...
         self.all_keys.extend([key for key in self.index if key not in self.all_keys])
+        self.device = device
 
     def __getitem__(self, key: str):
         # State dict gets priority
         if key in self.state_dict:
             return self.state_dict[key]
         weight_info = self.index[key]
+        if weight_info.get("safetensors_file") is not None:
+            from safetensors import safe_open
```
Comment (muellerzr) [inline src/accelerate/utils/offload.py:164]: We need a guard here for `is_safetensors_available` and raise an error if not :)

**Review by muellerzr — COMMENTED** [review]
Looks easy enough! Just one comment to make sure we add a `safetensors` check and it should be added to the test requirements most likely 😄

File: src/accelerate/utils/offload.py (hunk 1 above)
Comment (sgugger) [reply src/accelerate/utils/offload.py:164]: Indeed, will clean that up on Monday!

**Review by LysandreJik — APPROVED** [review]
:fire:

**Pull request merged**